tracer/src/Datadog.Trace/Configuration/SettingsManager.cs

```diff
@@ -0,0 +1,220 @@
+        /// static sources. To subscribe to updates to these settings, from code or remote config, call <see cref="SubscribeToChanges"/>.
+        /// </summary>
+        public ExporterSettings InitialExporterSettings { get; } = initialExporter;
+
+        /// <summary>
+        /// Subscribe to changes in <see cref="MutableSettings"/> and/or <see cref="ExporterSettings"/>.
+        /// Called whenever these settings change. If the settings have already changed when <see cref="SubscribeToChanges"/>
+        /// is called, <paramref name="callback"/> is invoked immediately with the latest configuration.
+        /// Also note that calling <see cref="SubscribeToChanges"/> twice with the same callback
+        /// will invoke the callback twice.
+        /// </summary>
+        /// <param name="callback">The method to invoke</param>
+        /// <returns>An <see cref="IDisposable"/> that should be disposed to unsubscribe</returns>
+        public IDisposable SubscribeToChanges(Action<SettingChanges> callback)
+        {
+            var subscription = new SettingChangeSubscription(this, callback);
+            lock (_subscribers)
+            {
+                _subscribers.Add(subscription);
+
+                if (Volatile.Read(ref _latest) is { } currentConfig)
+                {
+                    try
+                    {
+                        // If we already have updates, call this immediately
+                        callback(currentConfig);
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.Error(ex, "Error notifying subscriber of updated MutableSettings during subscribe");
+                    }
+                }
+            }
+
+            return subscription;
+        }
+
+        /// <summary>
+        /// Regenerate the application's new <see cref="MutableSettings"/> and <see cref="ExporterSettings"/>
+        /// based on runtime configuration sources.
+        /// </summary>
+        /// <param name="dynamicConfigSource">An <see cref="IConfigurationSource"/> for dynamic config via remote config</param>
+        /// <param name="manualSource">An <see cref="IConfigurationSource"/> for manual configuration (in code)</param>
+        /// <param name="centralTelemetry">The central <see cref="IConfigurationTelemetry"/> to report config telemetry updates to</param>
+        /// <returns>True if changes were detected and consumers were updated, false otherwise</returns>
+        public bool UpdateSettings(
+            IConfigurationSource dynamicConfigSource,
+            ManualInstrumentationConfigurationSourceBase manualSource,
+            IConfigurationTelemetry centralTelemetry)
+        {
+            if (BuildNewSettings(dynamicConfigSource, manualSource, centralTelemetry) is { } newSettings)
+            {
+                NotifySubscribers(newSettings);
```

Comment:
If two threads call `UpdateSettings()`, could there be a race condition where `NotifySubscribers()` is also called twice, since both can see the same `_latest` value in `BuildNewSettings()` below?

Ignore if you already considered this and decided that was acceptable.